File: doublelinked.py
class Node:
    __slots__="_element","_next","_prev"
    def __init__(self,element,next,prev):
        self._element=element
        self._next=next
        self._prev=prev


class Double:
    def __init__(self):
        self._head=None
        self._tail=None
        self._size=0

    def __len__(self):
        return self._size
    
    def isEmpty(self):
        return self._size==0
    

    def addLast(self,e):
        newest=Node(e,None,None)
        if self.isEmpty():
            self._head=newest
        else:
            self._tail._next=newest
            newest._prev=self._tail 
        self._tail=newest
        self._size+=1
    
    def display(self):
        p=self._head
        while p:
            print(p._element,end=".....>")
            p=p._next
        print()


    def addFirst(self,e):
        newest=Node(e,None,None)
        if self.isEmpty():
            self._head=newest
            self._tail=newest
        else:
            newest._next=self._head._next
            self._head._prev=newest
            self._head=newest
        self._size+=1

    def insertElement(self,e,pos):
        newest=Node(e,None,None)
        p=self._head
        i=1
        while i<(pos -1):
            p=p._next
            i+=1
        newest._next=p._next
        p._next._prev=newest
        p._next=newest
        newest._prev=p
        self._size+=1
    
    def delFirst(self):
        self._head=self._head._next
        self._head._prev=None
        self._size-=1
    
    def delEnd(self):
        i=0
        p=self._head
        while i<self._size:
            p=p._next
            i+=1
        self._tail=self._tail._prev
        self._tail._next=None
        self._size-=1


    def delAny(self,pos):
        i=0
        p=self._head
        while i<(pos-1):
            p=p._next
            i+=1
        p._next=p._next._next
        p._next._prev=p
        self._size-=1



            
d=Double()
d.addLast(1)
d.addLast(2)
d.addLast(3)
d.addLast(4)
d.addLast(5)
d.display()

File: double_linked.go
package main

import (
	"fmt"
	"strings"
)

type Node struct {
	value int
	next  *Node
	prev  *Node
}

type DoubleLinked struct {
	head *Node
	tail *Node
	len  int
}

func (d *DoubleLinked) addLast(val int) {
	newNode := &Node{value: val}
	if d.len == 0 {
		d.head = newNode
		d.head.prev = nil
	} else {
		d.tail.next = newNode
		newNode.prev = d.tail
	}
	d.tail = newNode
	d.len++
}

func (d *DoubleLinked) display() {
	ptr := d.head
	for i := 0; i < d.len; i++ {
		fmt.Println(ptr.value)
		ptr = ptr.next
	}
}

func (d *DoubleLinked) addFirst(val int) {
	node := &Node{value: val}
	if d.len == 0 {
		d.head = node
	} else {
		node.next = d.head
		d.head.prev = node
		d.head = node
	}
	d.len++
}

func (d *DoubleLinked) insertAny(val, pos int) {
	node := &Node{value: val}
	ptr := d.head
	for i := 1; i < pos-1; i++ {
		ptr = ptr.next
	}
	node.prev = ptr
	node.next = ptr.next
	ptr.next = node
	d.len++
}

func (d *DoubleLinked) delFirst() {
	if d.len == 0 {
		fmt.Println("list is empty")
	} else {
		d.head = d.head.next
		d.head.prev = nil
	}
	d.len--
}
func (d *DoubleLinked) delLast() {
	ptr := d.head
	for i := 1; i < d.len; i++ {
		ptr = ptr.next
	}
	ptr.next = nil
	fmt.Println(ptr)
	d.len--
}
func (d *DoubleLinked) delAny(pos int) {
	ptr := d.head
	for i := 1; i < pos-1; i++ {
		ptr = ptr.next
	}
	ptr.next = ptr.next.next
	ptr.next.prev = ptr
	d.len--
}

func main() {
	d := DoubleLinked{}
	d.addLast(1)
	d.addLast(2)
	d.addLast(3)
	d.addLast(4)
	d.display()
	fmt.Println(strings.Repeat("$", 30))
	d.addFirst(0)
	d.display()
	fmt.Println(strings.Repeat("$", 30))
	d.delFirst()
	d.display()
	fmt.Println(strings.Repeat("$", 30))
	d.insertAny(99, 3)
	d.delLast()
	d.display()
	fmt.Println(strings.Repeat("$", 30))
	d.delAny(3)
	d.display()
}
